### PR TITLE
Add Deneb state

### PR DESCRIPTION
**Pull request by terencechain**
This PR adds deneb beacon state native type. It also uses native type hash tree root for spec test

### CODE REVIEW

File: beacon-chain/state/state-native/getters_state.go [hunk 1]
```diff
@@ -338,3 +400,11 @@ func ProtobufBeaconStateCapella(s interface{}) (*ethpb.BeaconStateCapella, error
 	}
 	return pbState, nil
 }
+
+func ProtobufBeaconStateDeneb(s interface{}) (*ethpb.BeaconStateDeneb, error) {
+	pbState, ok := s.(*ethpb.BeaconStateDeneb)
+	if !ok {
+		return nil, errors.New("input is not type pb.BeaconStateCapella")
```
Comment (james-prysm) [inline beacon-chain/state/state-native/getters_state.go:407]: the error here should be StateDeneb I think

Comment (terencechain) [reply beacon-chain/state/state-native/getters_state.go]: a702121fb159098ee60700edaa70b8db08065699

File: beacon-chain/state/state-native/state_trie.go [hunk 2]
```diff
@@ -83,10 +83,19 @@ var capellaFields = append(
...
+	types.NextWithdrawalValidatorIndex,
+	types.HistoricalSummaries,
+)
+
 const (
 	phase0SharedFieldRefCount    = 10
 	altairSharedFieldRefCount    = 11
 	bellatrixSharedFieldRefCount = 12
+	denebSharedFieldRefCount     = 14
 	capellaSharedFieldRefCount   = 14
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/state/state-native/state_trie.go]: eb9fcd17aefde9dc0fdcde899c5a015687d72ec6

File: beacon-chain/state/state-native/state_fuzz_test.go [hunk 3]
```diff
@@ -292,3 +292,73 @@ func FuzzCapellaStateHashTreeRoot(f *testing.F) {
...
+			stateObj.Copy()
+
+			nativeState, err = coreState.ProcessSlots(context.Background(), nativeState, nativeState.Slot()+1)
+			assert.NoError(t, err)
+			nativeState.Copy()
+		}
+		assert.NoError(t, err)
+		// Perform a cold HTR calculation by initializing a new state.
+		innerState, ok := stateObj.ToProtoUnsafe().(*ethpb.BeaconStateDeneb)
+		assert.Equal(t, true, ok, "inner state is a not a beacon state bellatrix proto")
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/state/state-native/state_fuzz_test.go]: eb9fcd17aefde9dc0fdcde899c5a015687d72ec6

File: beacon-chain/state/state-native/hasher.go [hunk 4]
```diff
@@ -262,5 +264,31 @@ func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]b
...
+		nextWithdrawalValidatorIndexRoot := make([]byte, 32)
+		binary.LittleEndian.PutUint64(nextWithdrawalValidatorIndexRoot, uint64(state.nextWithdrawalValidatorIndex))
+		fieldRoots[types.NextWithdrawalValidatorIndex.RealPosition()] = nextWithdrawalValidatorIndexRoot
+
+		// Historical summary root.
+		historicalSummaryRoot, err := stateutil.HistoricalSummariesRoot(state.historicalSummaries)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not compute historical summary merkleization")
+		}
+		fieldRoots[types.HistoricalSummaries.RealPosition()] = historicalSummaryRoot[:]
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/state/state-native/hasher.go]: eb9fcd17aefde9dc0fdcde899c5a015687d72ec6

**Review by james-prysm — APPROVED** [review]

**Review by james-prysm — APPROVED** [review]

**Pull request merged**